### PR TITLE
Style branch selection

### PR DESCRIPTION
**Pull request by txsmith**
screenshot from 2016-05-14 19 29 12

### CODE REVIEW

**Comment by txsmith** [comment]
There is still one weird problem with the filtering. When a character from the currently selected branch is entered when the `paper-input` is empty, the `branch-selector` loses focus. The same happens when removing the last character from the `paper-input`

File: src/styles/app-theme.html [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
     --pr-status-color: #bbb;
     --anchor-color: #74B1FF;
+    --anchor-hover-color: red; /*#5EBFE8;*/
```
Comment (TimvdLippe) [inline src/styles/app-theme.html]: Should we uncomment this?

File: src/util/branch-tag.html [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        font-weight: 300;
+      }
+      :host(:hover) {
+        cursor: pointer;
+      }
+      list-item-label {
+        --tag-color: var(--anchor-color);
+      }
+      :host(:hover) list-item-label {
+        --tag-color: red;
```
Comment (TimvdLippe) [inline src/util/branch-tag.html]: Should probably be `var(--anchor-hover-color);`

**Comment by TimvdLippe** [comment]
Issue has been fixed :tada:

**Comment by eanker** [comment]
Toggle preview, toggles the description when no description is made and when there is a description it toggles the standard "Enter a description for your pull request. Markdown is supported.".

Furthermore when switching repo's, the title and the description previously filled in but not submitted are filled in.

**Comment by txsmith** [comment]
Your second point I know of but thought this was not very relevant to fix for the tuesday RITE session.

I cannot seem to reproduce your first point. When I enter some markdown and then click the toggle, it shows the rendered markdown, like this:

screenshot from 2016-05-15 19 00 39
screenshot from 2016-05-15 19 00 46

When the description is empty, the preview is also empty. But this is intended behavior.

**Comment by eanker** [comment]
Ah, I get it.
So, when you click toggle preview you can not edit anymore. 
I understood it incorrect, because of the fact that the title stays editable.

**Comment by txsmith** [comment]
Yes, when you go to an existing PR you can see what the non-editable version looks like

File: src/projects/project-create-pull-diff.html [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        margin-top: 0;
+      }
+    </style>
+    <h2>Describe your changes</h2>
+    <pull-request-diff editable data-route="diff" diff="[[diff]]"></pull-request-diff>
+  </template>
+  <script>
+    Polymer({
+      is: 'project-create-pull-diff',
+      behaviors: [
```
Comment (TimvdLippe) [inline src/projects/project-create-pull-diff.html]: This is duplicate with the file below. We should probably extract a Behavior for this.

File: src/projects/project-create-pull-request.html [hunk 4]
```diff
@@ -58,73 +49,82 @@
...
+        '_showDiff(diff)'
       ],
       _fetchBranchesAndCollaborators: function() {
         this.$.ajaxBranches.generateRequest();
         this.$.ajaxCollaborators.generateRequest();
       },
-      _fetchDiff: function() {
-        this.$.ajaxDiff.generateRequest();
+      _fetchDiff: function(baseBranch, headBranch) {
+        console.log('Fetching: ' + headBranch);
```
Comment (TimvdLippe) [inline src/projects/project-create-pull-request.html]: Left-over debugging statement.

File: src/pull-request/pull-request-chunk.html [hunk 5]
```diff
@@ -54,18 +58,57 @@
...
+        name="[[name]]"
+        ></markdown-input>
     </header>
-    <sortable-js group="chunk" draggable="pull-request-diff-highlight" on-remove="_diffRemoved">
-      <template is="dom-repeat" items="[[chunk.diff]]">
-        <pull-request-diff-highlight code="[[item.changes]]" file="[[item.file]]"></pull-request-diff-highlight>
-      </template>
-    </sortable-js>
-    <iron-icon icon="icons:arrow-downward" class="toggle-comments-icon" on-tap="_toggleComments"></iron-icon>
+    <template is="dom-if" if="[[editable]]">
```
Comment (TimvdLippe) [inline src/pull-request/pull-request-chunk.html]: Instead of a `dom-if` with copy-paste code, we should use the `disabled` property of `sortable-js`. E.g. `disabled="[[!editable]]"`

File: src/pull-request/pull-request-diff.html [hunk 6]
```diff
@@ -1,9 +1,17 @@
...
+<link rel="import" href="../bower_components/prism-element/prism-style-import.html">
 <link rel="import" href="../ajax/authenticated-ajax.html">
+<link rel="import" href="pull-request-chunk.html">
 
+<script src="../bower_components/prism/components/prism-diff.min.js"></script>
+// Match all lines that are not insertions or deletions
+Prism.languages.diff.other = /.+/m;
+// Additions and deletions can also be an empty line
+Prism.languages.diff.inserted = /^[+>].*$/m
+Prism.languages.diff.deleted = /^[-<].*$/m
```
Comment (TimvdLippe) [inline src/pull-request/pull-request-diff.html]: We should probably move these statements to a central location as high up the element chain as possible, because `<prism-highlighter>` captures all bubbled events. Bubbled events will bubble up all the way to `<rite-app>`

File: src/pull-request/pull-request-diff.html [hunk 7]
```diff
@@ -69,6 +79,10 @@
     Polymer({
       is: 'pull-request-diff',
       properties: {
+        editable: {
+          type: Boolean,
+          value: false
```
Comment (TimvdLippe) [inline src/pull-request/pull-request-diff.html]: `<pull-request-chunk>` have by default `editable = false`. So this should not be necessary.

File: src/pull-request/pull-request-diff.html [hunk 8]
```diff
@@ -101,7 +115,7 @@
       _removeChunk: function(e) {
         this.splice('chunks', e.model.index, 1);
       },
-      _parseDiffToChunks: (function() {
+      _parseDiffToChunks: function() {
```
Comment (TimvdLippe) [inline src/pull-request/pull-request-diff.html]: These parenthesis should not be removed. They create a scope for the regex declaration, so they are not leaking into global scope. Why did you remove them? (also for line 176)

File: src/projects/project-create-pull-diff.html [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        display: none;
+        background-color: var(--primary-background-color);
+        padding: 2em;
+      }
+      h2 {
+        margin-top: 0;
+      }
+    </style>
+    <h2>Describe your changes</h2>
+    <pull-request-diff editable data-route="diff" diff="[[diff]]"></pull-request-diff>
```
Comment (TimvdLippe) [inline src/projects/project-create-pull-diff.html]: This should not be editable right? Only on the actual pull requests pages we want to make them editable?

**Comment by TimvdLippe** [comment]
- [x] Initially it was not clear what I was supposed to do. Maybe instead of hiding the description and pull request diff, we could show a message similar to the No project selected. E.g. "Select a branch to merge".
- [x] When I click a branch tag, I expect the paper-input to be focused. (I might have broken this in my commit).
- [x] The input fields are not clearly separated. I have a hard time visually separating them. Maybe we should add some borders or different background-colors?
- [x] The shadow behind the pull-request-chunks is a bit too much imo. Maybe also switch to using borders?
- [x] "Enter a title" is used multiple times. Maybe we should differentiate between the pull request title and the title for a chunk. E.g. "Enter a pull request title" and "Enter a chunk title".
- [x] On hover on the + icon, I expect a tooltip (`title=""`) to show up with what it does. "Add a new chunk to this pull request".
- [x] Maybe we should also state to the user that you can drag the different chunks.
- [x] When I click "New comment" I get `VM131 polymer.html:2260 Uncaught TypeError: Cannot read property 'length' of undefined`. Probably we forgot to initialize an array of comments.
- [x] For some reason (I also observed this behavior in other branches) sometimes our `per-page` property of `authenticated-ajax` becomes undefined. Will have to test tomorrow more thoroughly what this is causing.
- [x] I think it is better to not hide the `-` icon on hover if you collapsed the chunk. When you did not collapse, I think showing on hover is just fine.
- [x] We should add a (visual, using css `::after`)down-arrow on the branch-selector to indicate it is a dropdown.
- [x] When switching from branches with different diff views, I get `pull-request-chunk.html:143 Uncaught TypeError: Cannot read property 'getElementsByClassName' of null`.

File: src/projects/project-create-pull-diff.html (hunk 9 above)
Comment (txsmith) [inline src/projects/project-create-pull-diff.html]: I think that would be very confusing for users. And what is the benefit of doing so? We needed to make the diff view editable anyway. Does the place in which it is editable matter from a technical standpoint?

Op 15 mei 2016 om 22:11 heeft Tim van der Lippe notifications@github.com het volgende geschreven:

In src/projects/project-create-pull-diff.html:

Comment (TimvdLippe) [inline src/projects/project-create-pull-diff.html]: Well, if a pull request has not been created yet you cant store the chunks. However, while I think of it we could push them when we create the pr and get a response back from github and only then store the chunks. In that sense they are delayed.

**Comment by TimvdLippe** [comment]
This PR is getting really big, so I would say we merge this sooner rather than later. Most of the issues are fixed. Let's fixed the last couple of pionts in a different PR?

**Comment by eanker** [comment]
Merge
